File: src/claude_config/egress_broker/wire.py
"""Wire protocol for the egress broker UDS.

Each request/response is a single length-prefixed JSON frame:

    +--------+--------------------------------+
    | u32 BE | JSON bytes (max 100 MiB)       |
    +--------+--------------------------------+

The reference implementation does NOT stream bodies — a request is read
whole, the upstream call is made whole, the response is returned whole.
This is adequate for the request shapes the sandbox makes (chat
completions are kilobytes; the broker is not in the data-plane for
large transfers). Production Go implementation may add streaming.

Request JSON shape (sandbox -> broker)::

    {
      "alias":   "anthropic-api",
      "method":  "POST",
      "path":    "/v1/messages",
      "headers": {"content-type": ["application/json"], ...},
      "body_b64": "<base64-encoded body, may be empty string>"
    }

Response JSON shape (broker -> sandbox)::

    {
      "status":  200,
      "headers": {"content-type": ["application/json"], ...},
      "body_b64": "<base64-encoded body>"
    }

Error response (broker -> sandbox)::

    {
      "error":   "policy-denied" | "upstream-failed" | "bad-request",
      "message": "human-readable detail"
    }
"""

from __future__ import annotations

import base64
import json
import socket
import struct
from dataclasses import dataclass
from typing import Any

MAX_FRAME_BYTES = 100 * 1024 * 1024
_LEN_STRUCT = struct.Struct("!I")


class WireError(ValueError):
    """A wire-format violation."""


@dataclass(frozen=True)
class Request:
    """A parsed sandbox request."""

    alias: str
    method: str
    path: str
    headers: dict[str, list[str]]
    body: bytes

    @classmethod
    def from_json(cls, obj: dict[str, Any]) -> Request:
        """Validate-and-construct from a parsed JSON object."""
        try:
            alias = obj["alias"]
            method = obj["method"]
            path = obj["path"]
            headers_raw = obj["headers"]
            body_b64 = obj["body_b64"]
        except KeyError as e:
            raise WireError(f"missing required field: {e.args[0]}") from None
        if not isinstance(alias, str) or not alias:
            raise WireError("'alias' must be a non-empty string")
        if not isinstance(method, str) or method != method.upper():
            raise WireError("'method' must be uppercase ASCII")
        if not isinstance(path, str) or not path.startswith("/"):
            raise WireError("'path' must be a string starting with '/'")
        if not isinstance(headers_raw, dict):
            raise WireError("'headers' must be an object")
        headers: dict[str, list[str]] = {}
        for k, v in headers_raw.items():
            if not isinstance(k, str):
                raise WireError("header names must be strings")
            if not isinstance(v, list) or not all(isinstance(s, str) for s in v):
                raise WireError(f"header values for {k!r} must be a list of strings")
            headers[k] = list(v)
        if not isinstance(body_b64, str):
            raise WireError("'body_b64' must be a string")
        try:
            body = base64.b64decode(body_b64, validate=True) if body_b64 else b""
        except ValueError as e:
            raise WireError(f"'body_b64' is not valid base64: {e}") from None
        return cls(alias=alias, method=method, path=path, headers=headers, body=body)


@dataclass(frozen=True)
class Response:
    """A broker response to forward back to the sandbox."""

    status: int
    headers: dict[str, list[str]]
    body: bytes

    def to_json(self) -> dict[str, Any]:
        """Serialize to the wire JSON shape."""
        return {
            "status": self.status,
            "headers": self.headers,
            "body_b64": base64.b64encode(self.body).decode("ascii"),
        }


def encode_frame(payload: dict[str, Any]) -> bytes:
    """Encode a JSON payload as a length-prefixed frame."""
    body = json.dumps(payload, separators=(",", ":")).encode("utf-8")
    if len(body) > MAX_FRAME_BYTES:
        raise WireError(f"frame body {len(body)} exceeds MAX_FRAME_BYTES={MAX_FRAME_BYTES}")
    return _LEN_STRUCT.pack(len(body)) + body


def read_frame(sock: socket.socket) -> dict[str, Any]:
    """Read one length-prefixed JSON frame from ``sock``.

    Raises:
        WireError: On short read, oversized frame, or invalid JSON.
    """
    header = _recv_exact(sock, _LEN_STRUCT.size)
    (length,) = _LEN_STRUCT.unpack(header)
    if length > MAX_FRAME_BYTES:
        raise WireError(f"declared frame length {length} exceeds MAX_FRAME_BYTES")
    body = _recv_exact(sock, length)
    try:
        obj = json.loads(body)
    except json.JSONDecodeError as e:
        raise WireError(f"frame body is not valid JSON: {e}") from None
    if not isinstance(obj, dict):
        raise WireError("frame body must be a JSON object")
    return obj


def _recv_exact(sock: socket.socket, n: int) -> bytes:
    """Read exactly ``n`` bytes from ``sock`` or raise WireError."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise WireError(f"short read: wanted {n} bytes, got {len(buf)} before EOF")
        buf.extend(chunk)
    return bytes(buf)

File: sandbox/broker/internal/wire/wire.go
// Package wire implements the egress broker's UDS framing protocol.
//
// Each request/response is a single length-prefixed JSON frame:
//
//	+--------+--------------------------------+
//	| u32 BE | JSON bytes (max 100 MiB)       |
//	+--------+--------------------------------+
//
// Bodies are carried base64-encoded in the JSON payload; no streaming.
// Parity reference: src/claude_config/egress_broker/wire.py.
package wire

import (
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

const MaxFrameBytes = 100 * 1024 * 1024

// ErrWire wraps any wire-format violation.
type ErrWire struct{ Msg string }

func (e *ErrWire) Error() string { return e.Msg }

func wireErr(format string, args ...any) error {
	return &ErrWire{Msg: fmt.Sprintf(format, args...)}
}

// Request is a parsed sandbox request.
type Request struct {
	Alias   string
	Method  string
	Path    string
	Headers map[string][]string
	Body    []byte
}

// Response is a broker response to forward back to the sandbox.
type Response struct {
	Status  int
	Headers map[string][]string
	Body    []byte
}

// requestJSON / responseJSON / errorJSON are the wire envelopes.
type requestJSON struct {
	Alias   string              `json:"alias"`
	Method  string              `json:"method"`
	Path    string              `json:"path"`
	Headers map[string][]string `json:"headers"`
	BodyB64 string              `json:"body_b64"`
}

type responseJSON struct {
	Status  int                 `json:"status"`
	Headers map[string][]string `json:"headers"`
	BodyB64 string              `json:"body_b64"`
}

// ErrorEnvelope is the wire shape for negative responses.
type ErrorEnvelope struct {
	Code    string `json:"error"`
	Message string `json:"message"`
}

// ParseRequest validates a raw JSON object and returns a Request.
func ParseRequest(raw map[string]json.RawMessage) (*Request, error) {
	var rj requestJSON
	// Re-marshal then unmarshal into the typed struct so error messages
	// stay close to the Python reference's "missing required field" form.
	for _, k := range []string{"alias", "method", "path", "headers", "body_b64"} {
		if _, ok := raw[k]; !ok {
			return nil, wireErr("missing required field: %s", k)
		}
	}
	buf, _ := json.Marshal(raw)
	if err := json.Unmarshal(buf, &rj); err != nil {
		return nil, wireErr("invalid request shape: %s", err)
	}
	if rj.Alias == "" {
		return nil, wireErr("'alias' must be a non-empty string")
	}
	if rj.Method == "" || rj.Method != strings.ToUpper(rj.Method) {
		return nil, wireErr("'method' must be uppercase ASCII")
	}
	if !strings.HasPrefix(rj.Path, "/") {
		return nil, wireErr("'path' must be a string starting with '/'")
	}
	if rj.Headers == nil {
		return nil, wireErr("'headers' must be an object")
	}
	var body []byte
	if rj.BodyB64 != "" {
		b, err := base64.StdEncoding.DecodeString(rj.BodyB64)
		if err != nil {
			return nil, wireErr("'body_b64' is not valid base64: %s", err)
		}
		body = b
	}
	hdrs := make(map[string][]string, len(rj.Headers))
	for k, v := range rj.Headers {
		hdrs[k] = append([]string(nil), v...)
	}
	return &Request{
		Alias:   rj.Alias,
		Method:  rj.Method,
		Path:    rj.Path,
		Headers: hdrs,
		Body:    body,
	}, nil
}

// EncodeResponse serializes a Response into a wire-ready JSON payload.
func EncodeResponse(r *Response) map[string]any {
	hdrs := r.Headers
	if hdrs == nil {
		hdrs = map[string][]string{}
	}
	return map[string]any{
		"status":   r.Status,
		"headers":  hdrs,
		"body_b64": base64.StdEncoding.EncodeToString(r.Body),
	}
}

// EncodeError serializes an error envelope into a wire payload.
func EncodeError(code, message string) map[string]any {
	return map[string]any{"error": code, "message": message}
}

// EncodeFrame serializes a payload as a length-prefixed JSON frame.
func EncodeFrame(payload any) ([]byte, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, wireErr("json marshal failed: %s", err)
	}
	if len(body) > MaxFrameBytes {
		return nil, wireErr("frame body %d exceeds MAX_FRAME_BYTES=%d", len(body), MaxFrameBytes)
	}
	out := make([]byte, 4+len(body))
	binary.BigEndian.PutUint32(out[:4], uint32(len(body)))
	copy(out[4:], body)
	return out, nil
}

// ReadFrame reads one length-prefixed JSON frame from r as a raw map.
func ReadFrame(r io.Reader) (map[string]json.RawMessage, error) {
	var hdr [4]byte
	if _, err := io.ReadFull(r, hdr[:]); err != nil {
		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
			return nil, wireErr("short read: wanted 4 bytes, hit EOF")
		}
		return nil, wireErr("frame header read failed: %s", err)
	}
	length := binary.BigEndian.Uint32(hdr[:])
	if int(length) > MaxFrameBytes {
		return nil, wireErr("declared frame length %d exceeds MAX_FRAME_BYTES", length)
	}
	body := make([]byte, length)
	if _, err := io.ReadFull(r, body); err != nil {
		return nil, wireErr("short read: wanted %d body bytes: %s", length, err)
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(body, &obj); err != nil {
		return nil, wireErr("frame body is not valid JSON object: %s", err)
	}
	if obj == nil {
		return nil, wireErr("frame body must be a JSON object")
	}
	return obj, nil
}
